File: Java/jobspls/algorithms/FrogRiverJump.java
package jobspls.algorithms;

import java.util.Arrays;
import java.util.OptionalInt;
import java.util.TreeSet;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class FrogRiverJump {
    public static OptionalInt frog_jump(int jumps, int[] leaves) {
        var required = new TreeSet<Integer>(IntStream
            .range(1, jumps + 1)
            .boxed()
            .collect(Collectors.toList()));

        var time = 0;
        for (var leaf : leaves) {
            required.remove(leaf);
            if (required.isEmpty()) {
                return OptionalInt.of(time);
            }
            time++;
        }

        return OptionalInt.empty();
    }
}

File: Go/jobspls/algorithms/frog_river_jump.go
package algorithms

// When can froggy jump to the other side of the river?
//
// Froggy needs 1 to `jumps` leaves.
// `leaves` is a slice of flowing leaves where each index is a second of time
func FrogRiverJump(jumps int, leaves []int) (int, bool) {
	required := make(map[int]struct{}, jumps)
	for i := 1; i <= jumps; i++ {
		required[i] = struct{}{}
	}

	for time, leaf := range leaves {
		delete(required, leaf)
		if len(required) == 0 {
			return time, true
		}
	}

	var nope int
	return nope, false
}
